### PR TITLE
feat(tools): add transparent async tool support (fixes #334)

### LINKED ISSUE #334: Async tool support
Are async functions currently unsupported, or is there a workaround for tools that rely on asynchronous behaviour?

For example, I have a tool that retrieves data asynchronously from various web resources, and the agent needs to await these responses:
```
@tool
async def retrieve_datasets(query: str, retriever: RepoRetriever) -> dict:
    """
    This tool can be used to retrieve metadata from different repositories.

    Note this is a async function and needs to be awaited.

    Args:
        query: The query string to search for.
        retriever: An instance of the RepoRetriever class.
    """
    response = await(retriever.query_all_repos(query))

    top_5_results = response.get("results", [])[:5]
    return top_5_results
```

Attempting to use this tool results in the following error:
 ```
─ Executing this code: ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  precipitation_data = await retrieve_datasets(query="precipitation data", retriever=retriever)
  print(precipitation_data)
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Code execution failed at line 'precipitation_data = await retrieve_datasets(query="precipitation data", retriever=retriever)' because of the following error:
Await is not supported.
```
Is there a recommended approach to handle asynchronous tools in such scenarios? Any guidance would be greatly appreciated.

### PR DESCRIPTION
**Pull request by giulio-leone**
## Problem
smolagents does not support async tools — defining an async `forward()` method returns a coroutine object instead of the actual result. This is a significant limitation for tools that need async I/O (database queries, API calls, file operations).

Related: #145 (61 reactions), #334 (8 reactions)

## Solution
Add transparent async detection and execution in the `Tool.__call__()` method:

1. After calling `self.forward(*args, **kwargs)`, check if the result is awaitable via `inspect.isawaitable()`
2. If no event loop is running: use `asyncio.run()` to execute it
3. If inside a running event loop: use a thread pool to avoid nested event loop issues

The `@tool` decorator's AST parsing also now handles `AsyncFunctionDef` nodes, so `@tool`-decorated async functions work correctly.

This is **fully backward compatible** — sync tools work exactly as before. Async tools now 'just work' without any configuration.

### Changes
- `src/smolagents/tools.py`: Added `_run_async()` helper, async handling in `Tool.__call__()`, `AsyncFunctionDef` support in `@tool` decorator
- `tests/test_async_tools.py`: 10 new tests covering subclass tools, decorated tools, event loop contexts, and various return types

## Testing
- 10 new tests for async tool execution all pass
- All 70 existing non-MCP tool tests continue to pass (3 MCP integration tests were already failing pre-change due to missing `python` binary)

Fixes #334
Related: #145

### CODE REVIEW

File: src/smolagents/tools.py [hunk 1]
```diff
@@ -1115,8 +1136,10 @@ def wrapped_function(*args, **kwargs):
     # - Remove the tool decorator and function definition line
     lines = tool_source.splitlines()
     tree = ast.parse(tool_source)
-    #   - Find function definition
-    func_node = next((node for node in ast.walk(tree) if isinstance(node, ast.FunctionDef)), None)
+    #   - Find function definition (sync or async)
+    func_node = next(
+        (node for node in ast.walk(tree) if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef))), None
+    )
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply src/smolagents/tools.py:1180]: Fixed in f745f04: The `@tool` decorator now checks `inspect.iscoroutinefunction(tool_function)` and generates `async def forward` for async functions.

File: src/smolagents/tools.py [hunk 2]
```diff
@@ -95,6 +96,23 @@ def new_init(self, *args, **kwargs):
...
+    except RuntimeError:
+        loop = None
+    if loop is not None and loop.is_running():
+        # Inside a running event loop — run in a separate thread to avoid blocking
+        import concurrent.futures
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=1) as pool:
+            return pool.submit(asyncio.run, coro).result()
+    else:
+        return asyncio.run(coro)
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply src/smolagents/tools.py]: Fixed in f745f04: `_run_async` now wraps non-coroutine awaitables via `inspect.iscoroutine()` check before passing to `asyncio.run()`.

File: src/smolagents/tools.py [hunk 3]
```diff
@@ -95,6 +96,23 @@ def new_init(self, *args, **kwargs):
...
+        loop = asyncio.get_running_loop()
+    except RuntimeError:
+        loop = None
+    if loop is not None and loop.is_running():
+        # Inside a running event loop — run in a separate thread to avoid blocking
+        import concurrent.futures
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=1) as pool:
+            return pool.submit(asyncio.run, coro).result()
+    else:
```
[the remark this answers is not in the document]
Comment (giulio-leone) [reply src/smolagents/tools.py:150]: Fixed in f745f04: Added comprehensive docstring to `_run_async` explaining that `.result()` blocking is intentional — the `Tool.__call__` contract is synchronous, and the thread-pool offload keeps the event loop unblocked.

File: src/smolagents/tools.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply src/smolagents/tools.py:1180]: Fixed in f745f04: Added `TestAsyncToolSerialization` class with `test_async_subclass_round_trip` and `test_async_decorated_tool_round_trip` — both verify `to_dict()` / `from_dict()` round-tripping.

File: src/smolagents/tools.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply src/smolagents/tools.py]: This is already handled. Lines 115–118 wrap non-coroutine awaitables in a thin coroutine before passing to `asyncio.run()`:

```python
if not inspect.iscoroutine(coro):
    async def _wrap():
        return await coro
    coro = _wrap()
```

So if `forward()` returns an arbitrary awaitable (e.g. an object with `__await__`), it gets promoted to a proper coroutine first. The dual-path approach then covers both cases: `asyncio.run()` when no loop is running, and `ThreadPoolExecutor` + `asyncio.run()` when a loop already exists.

File: src/smolagents/tools.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply src/smolagents/tools.py:1180]: Acknowledged — the current tests cover core async execution paths (basic invocation, the dual event-loop branches, and the `@tool` decorator with async functions). Serialization round-trip tests (`to_dict` / `from_dict`) for async `@tool`-decorated tools would strengthen coverage and are noted for a follow-up. Event loop interaction edge cases (e.g. nested loops, loop policy variations) are harder to test deterministically across platforms.

[the remark this answers is not in the document]
Comment (giulio-leone) [reply src/smolagents/tools.py:1180]: This is already handled. The source generation at lines 1184–1187 checks `inspect.iscoroutinefunction(tool_function)` and conditionally emits `async def forward`:

```python
if inspect.iscoroutinefunction(tool_function):
    forward_method_source = f"async def forward{new_sig}:\n{tool_source_body}"
else:
    forward_method_source = f"def forward{new_sig}:\n{tool_source_body}"
```

The `AsyncFunctionDef` detection in the AST walk (the line this comment is on) is for correctly locating the function node regardless of whether it's sync or async — the actual `__source__` generation downstream preserves the async nature.

File: src/smolagents/tools.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (giulio-leone) [reply src/smolagents/tools.py:150]: This is intentional by design — `Tool.__call__` has a synchronous contract, so callers expect a return value, not an awaitable.

The `ThreadPoolExecutor` offloads `asyncio.run(coro)` to a **separate thread** with its own event loop, so the caller's event loop is not directly stalled. The `.result()` call blocks only the calling thread while waiting for the worker thread to finish, which is the standard pattern for bridging sync→async when a loop already exists (see `asgiref.sync_to_async` and similar utilities).

This is documented in the function docstring: _"The subsequent `.result()` call blocks the calling thread, which is acceptable because the `Tool.__call__` contract is synchronous."_
